lib/rules/prefer-arrow-callback.js

```diff
@@ -246,7 +249,37 @@ module.exports = {
                     !scopeInfo.super &&
                     !scopeInfo.meta
                 ) {
-                    context.report(node, "Unexpected function expression.");
+                    context.report({
+                        node,
+                        message: "Unexpected function expression.",
+                        fix(fixer) {
+
+                            // FIXME: Is there a better way to detect parser support for arrow functions?
+                            if (!context.parserOptions || !context.parserOptions.ecmaVersion || context.parserOptions.ecmaVersion < 6) {
+
+                                // If ES6 parsing is not enabled, creating arrow functions will cause syntax errors, so don't do that.
+                                return null;
+                            }
+
+                            if (!callbackInfo.isLexicalThis && scopeInfo.this) {
+
+                                // If the callback function does not have .bind(this) and contains a reference to `this`, there
+                                // is no way to determine what `this` should be, so don't perform any fixes.
+                                return null;
+                            }
+
+                            const params = "(" + node.params.map(param => sourceCode.getText(param)).join(", ") + ")";
```

Comment:
That works, thanks. :+1:

I noticed that when I tried out the example you gave in Chrome's console, I got a SyntaxError:

```js
foo(function({a, b = 1, c: [d, a]}) { })
// --> Uncaught SyntaxError: Duplicate parameter name not allowed in this context
```

But when I tried it in the ESLint demo, I didn't get a SyntaxError:

![](https://i.gyazo.com/a123f2cd00329a8a3f5ef3e7ec7de894.png)

Looking at [the spec](http://www.ecma-international.org/ecma-262/7.0/#sec-function-definitions-static-semantics-early-errors), it seems like Chrome's behavior is correct:

> FormalParameters:FormalParameterList

> * It is a Syntax Error if IsSimpleParameterList of *FormalParameterList* is **false** and BoundNames of *FormalParameterList* contains any duplicate elements.